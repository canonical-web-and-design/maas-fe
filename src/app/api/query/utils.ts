import { useMemo } from "react";

import type { UseQueryResult } from "@tanstack/react-query";

type QueryHook<T> = () => UseQueryResult<T[], unknown>;

export const useItemsCount = <T>(useItems: QueryHook<T>) => {
  const { data } = useItems();
  return useMemo(() => data?.length ?? 0, [data]);
};
