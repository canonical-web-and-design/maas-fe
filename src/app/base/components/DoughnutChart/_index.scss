@mixin DoughnutChart {
  .doughnut-chart__tooltip {
    display: block;
  }

  .doughnut-chart__tooltip > span {
    // Override the tooltip wrapper.
    display: block !important;
  }

  .doughnut-chart__chart {
    // Restrict hover areas to the strokes.
    pointer-events: stroke;
  }

  .doughnut-chart__label {
    // Center the text in the circle.
    dominant-baseline: middle;
    text-anchor: middle;
  }

  .doughnut-chart__segment {
    fill: transparent;
    // Animate stroke size changes on hover.
    transition: stroke-width 0.3s ease;
  }
}
