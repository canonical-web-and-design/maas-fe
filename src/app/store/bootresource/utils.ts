import type { BootResource } from "./types";

export const splitResourceName = (
  name: BootResource["name"]
): { os: string; release: string } => {
  const split = name.split("/");
  return split.length === 2
    ? { os: split[0], release: split[1] }
    : { os: "", release: "" };
};

export const splitImageName = (
  name: string
): { os: string; arch: string; subArch: string; release: string } => {
  const split = name.split("/");
  return split.length === 4
    ? { os: split[0], arch: split[1], subArch: split[2], release: split[3] }
    : { os: "", arch: "", subArch: "", release: "" };
};
