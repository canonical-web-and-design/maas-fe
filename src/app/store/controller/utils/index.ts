export {
  isControllerDetails,
  isRack,
  isRegionAndRack,
  isRegion,
} from "./common";
export { FilterControllers, getControllerValue } from "./search";
