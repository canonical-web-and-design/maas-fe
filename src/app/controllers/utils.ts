import { ControllerSidePanelViews } from "./constants";

import type { SidePanelContent } from "app/base/side-panel-context";
import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param sidePanelContent - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  sidePanelContent: SidePanelContent | null
): string => {
  if (sidePanelContent) {
    const [, name] = sidePanelContent.view;
    switch (name) {
      case ControllerSidePanelViews.ADD_CONTROLLER[1]:
        return "Add controller";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
