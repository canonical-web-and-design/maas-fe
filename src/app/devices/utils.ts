import { DeviceHeaderViews } from "./constants";
import type { DeviceSidePanelContent } from "./types";

import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param sidePanelContent - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  sidePanelContent: DeviceSidePanelContent | null
): string => {
  if (sidePanelContent) {
    const [, name] = sidePanelContent.view;
    switch (name) {
      case DeviceHeaderViews.ADD_DEVICE[1]:
        return "Add device";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
