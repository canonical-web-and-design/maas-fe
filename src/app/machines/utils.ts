import { MachineHeaderViews } from "./constants";
import type { MachineSidePanelContent } from "./types";

import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param sidePanelContent - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  sidePanelContent: MachineSidePanelContent | null
): string => {
  if (sidePanelContent) {
    const [, name] = sidePanelContent.view;
    switch (name) {
      case MachineHeaderViews.ADD_CHASSIS[1]:
        return "Add chassis";
      case MachineHeaderViews.ADD_MACHINE[1]:
        return "Add machine";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
