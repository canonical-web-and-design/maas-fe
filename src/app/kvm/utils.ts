import { KVMHeaderViews } from "./constants";
import type {
  KVMSidePanelContent,
  KVMStoragePoolResource,
  KVMStoragePoolResources,
} from "./types";

import type { MachineSidePanelContent } from "app/machines/types";
import { getHeaderTitle as getMachineHeaderTitle } from "app/machines/utils";
import type { Pod } from "app/store/pod/types";
import { formatBytes } from "app/utils";

/**
 * Returns a string with the formatted byte value and unit, e.g 1024 => "1KiB"
 *
 * @param memory - the memory in bytes
 * @returns formatted memory string with value and unit
 */
export const memoryWithUnit = (memory: number): string => {
  const formatted = formatBytes(memory, "B", { binary: true });
  return `${formatted.value}${formatted.unit}`;
};

/**
 * Get header title depending on header content.
 * @param sidePanelContent - The currently selected header content.
 * @returns Header title.
 */
export const getFormTitle = (sidePanelContent: KVMSidePanelContent): string => {
  switch (sidePanelContent.view) {
    case KVMHeaderViews.ADD_LXD_HOST:
      return "Add LXD host";
    case KVMHeaderViews.ADD_VIRSH_HOST:
      return "Add Virsh host";
    case KVMHeaderViews.COMPOSE_VM:
      return "Compose";
    case KVMHeaderViews.DELETE_KVM:
      return "Delete";
    case KVMHeaderViews.REFRESH_KVM:
      return "Refresh";
    default:
      // We need to explicitly cast sidePanelContent here - TypeScript doesn't
      // seem to be able to infer remaining object tuple values as with string
      // values.
      // https://github.com/canonical/maas-ui/issues/3040
      const machineSidePanelContent =
        sidePanelContent as MachineSidePanelContent;
      return getMachineHeaderTitle("", machineSidePanelContent);
  }
};

/**
 * Calculate the amount of free storage in a storage pool.
 * @param resource - The storage pool resource to calculate free storage.
 * @returns Free storage in pool.
 */
export const calcFreePoolStorage = (resource: KVMStoragePoolResource): number =>
  resource.total - resource.allocated_other - resource.allocated_tracked;

/**
 * Convert a pod or cluster's storage pool resources object into a sorted array.
 * @param pools - The pod or cluster's storage pool resources object.
 * @param defaultPoolId - the default pool id of the pod.
 * @returns a sorted list of storage pools in the pod or cluster.
 */
export const getSortedPoolsArray = (
  pools: KVMStoragePoolResources,
  defaultPoolId?: Pod["default_storage_pool"]
): [name: string, resource: KVMStoragePoolResource][] => {
  const poolsArray = Object.entries<KVMStoragePoolResource>(pools);

  return poolsArray.sort(([nameA, dataA], [nameB, dataB]) => {
    if (defaultPoolId && "id" in dataA && "id" in dataB) {
      // Pools in pods will have an id. For this case we sort by default first
      // (as defined in pod.default_storage_pool) then by id.
      if (
        dataA.id === defaultPoolId ||
        (dataB.id !== defaultPoolId && dataA.id < dataB.id)
      ) {
        return -1;
      } else if (dataB.id === defaultPoolId || dataA.id > dataB.id) {
        return 1;
      }
      return 0;
    }

    // Pools in clusters do not have a single id, as they can span multiple
    // pods. For this case we just sort by name;
    if (nameA < nameB) {
      return -1;
    } else if (nameA > nameB) {
      return 1;
    }
    return 0;
  });
};
