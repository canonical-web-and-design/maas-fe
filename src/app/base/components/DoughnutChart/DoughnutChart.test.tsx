import DoughnutChart, { TestIds } from "./DoughnutChart";

import { render, screen } from "testing/utils";

describe("DoughnutChart", () => {
  it("renders", () => {
    const { container } = render(
      <DoughnutChart
        label="200GB"
        segmentHoverWidth={20}
        segmentWidth={15}
        segments={[
          {
            color: "#06C",
            tooltip: "Allocated",
            value: 20,
          },
          {
            color: "#cce0f5",
            tooltip: "Free",
            value: 5,
          },
        ]}
        size={96}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("can render with a label", () => {
    render(
      <DoughnutChart
        label="Label!"
        segmentHoverWidth={20}
        segmentWidth={15}
        segments={[
          {
            color: "#06C",
            value: 20,
          },
          {
            color: "#cce0f5",
            value: 5,
          },
        ]}
        size={96}
      />
    );

    expect(screen.getByTestId(TestIds.Label)).toBeInTheDocument();
  });

  it("can render without a label", () => {
    render(
      <DoughnutChart
        segmentHoverWidth={20}
        segmentWidth={15}
        segments={[
          {
            color: "#06C",
            value: 20,
          },
          {
            color: "#cce0f5",
            value: 5,
          },
        ]}
        size={96}
      />
    );

    expect(screen.queryByTestId(TestIds.Label)).not.toBeInTheDocument();
  });
});
