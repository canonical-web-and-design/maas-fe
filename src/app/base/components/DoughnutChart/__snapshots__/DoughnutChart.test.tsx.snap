// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoughnutChart renders 1`] = `
<div>
  <div
    class="doughnut-chart"
    style="max-width: 106px;"
  >
    <span
      class="doughnut-chart__tooltip"
    >
      <style>
        #doughnut-chart-mock-redux-js-nanoid-1 .doughnut-chart__segment:hover {
          stroke-width: 25 !important;
        }
      </style>
      <svg
        class="doughnut-chart__chart"
        id="doughnut-chart-mock-redux-js-nanoid-1"
        viewBox="0 0 106 106"
      >
        <mask
          id="myMask"
        >
          <rect
            fill="white"
            height="106"
            width="106"
            x="0"
            y="0"
          />
          <circle
            cx="53"
            cy="53"
            fill="black"
            r="33"
          />
        </mask>
        <g
          mask="url(#myMask)"
        >
          <rect
            fill="transparent"
            height="106"
            width="106"
            x="0"
            y="0"
          />
          <g>
            <circle
              class="doughnut-chart__segment"
              cx="28"
              cy="53"
              data-testid="segment"
              r="40.5"
              style="stroke: #06C; stroke-width: 15; stroke-dasharray: 0 0.00 203.20 50.80;"
              transform="rotate(-90 40.5,40.5)"
            />
            <circle
              class="doughnut-chart__segment"
              cx="28"
              cy="53"
              data-testid="segment"
              r="40.5"
              style="stroke: #cce0f5; stroke-width: 15; stroke-dasharray: 0 203.20 50.80 -0.00;"
              transform="rotate(-90 40.5,40.5)"
            />
          </g>
        </g>
        <text
          x="53"
          y="53"
        >
          <tspan
            class="doughnut-chart__label"
            data-testid="label"
          >
            200GB
          </tspan>
        </text>
      </svg>
    </span>
  </div>
</div>
`;
